shell/platform/linux/fl_platform_plugin.cc

```diff
@@ -140,6 +153,124 @@ static FlMethodResponse* clipboard_has_strings_async(
   return nullptr;
 }
 
+// Get the exit response from a System.requestAppExit method call.
+static gchar* get_exit_response(FlMethodResponse* response) {
+  if (response == nullptr) {
+    return nullptr;
+  }
+
+  g_autoptr(GError) error = nullptr;
+  g_autoptr(FlValue) result = fl_method_response_get_result(response, &error);
+  if (result == nullptr) {
+    g_warning("Error returned from System.requestAppExit: %s", error->message);
+    return nullptr;
+  }
+  if (fl_value_get_type(result) != FL_VALUE_TYPE_MAP) {
+    g_warning("System.requestAppExit result argument map missing or malformed");
+    return nullptr;
+  }
+
+  FlValue* response_value = fl_value_lookup_string(result, kExitResponseKey);
+  if (fl_value_get_type(response_value) != FL_VALUE_TYPE_STRING) {
+    g_warning("Invalid response from System.requestAppExit");
+    return nullptr;
+  }
+  return g_strdup(fl_value_get_string(response_value));
+}
+
+// Handle response of System.requestAppExit.
+static void request_app_exit_response_cb(GObject* object,
+                                         GAsyncResult* result,
+                                         gpointer user_data) {
+  FlPlatformPlugin* self = FL_PLATFORM_PLUGIN(user_data);
+
+  g_autoptr(GError) error = nullptr;
+  g_autoptr(FlMethodResponse) method_response =
+      fl_method_channel_invoke_method_finish(FL_METHOD_CHANNEL(object), result,
+                                             &error);
+  g_autofree gchar* exit_response = nullptr;
+  if (method_response == nullptr) {
+    if (g_error_matches(error, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
+      return;
+    }
+    g_warning("Failed to complete System.requestAppExit: %s", error->message);
+  } else {
+    exit_response = get_exit_response(method_response);
+  }
+  if (exit_response == nullptr) {
+    exit_response = g_strdup(kExitResponseCancel);
+  }
+
+  if (g_str_equal(exit_response, kExitResponseExit)) {
+    GApplication* app = g_application_get_default();
+    if (app == nullptr) {
+      g_warning("Unable to quit, no GApplication");
+    } else {
+      g_application_quit(app);
```

Comment:
BTW, I tried this with the Framework code, and even if it responds "cancel" and never executes this line, the application exits once it returns from this callback. I'm confused as to why, since it definitely isn't calling `g_application_quit` in that case (or in the other instance in this file). Is there another reason it might exit?

You can try it out with the framework changes too: patch in https://github.com/flutter/flutter/pull/121378 and run `examples/api/lib/widgets/application_lifecycle/application_lifecycle.0.dart`.  This works with the macOS version, so I'm pretty sure it's sending the right communication.

Reply:
Also, it doesn't seem to go through `fl_platform_plugin_request_app_exit` when I click on the window "X" to close it. I would expect the app to get a chance to ask the user a question before exiting in that case.

Reply:
Hmm. Seems that it's a SEGV after that function returns, for some reason it's failing here:

```
Thread 1 "flutter_api_sam" received signal SIGSEGV, Segmentation fault.
malloc_consolidate (av=av@entry=0x7f1b83b12c60 <main_arena>) at ./malloc/malloc.c:4749
4749	./malloc/malloc.c: No such file or directory.

#0  malloc_consolidate (av=av@entry=0x7f1b83b12c60 <main_arena>) at ./malloc/malloc.c:4749
#1  0x00007f1b839d76b8 in _int_malloc (av=av@entry=0x7f1b83b12c60 <main_arena>, bytes=bytes@entry=3256) at ./malloc/malloc.c:3961
#2  0x00007f1b839d8799 in __GI___libc_malloc (bytes=3256) at ./malloc/malloc.c:3323
#3  0x00007f1b873eec3d in operator new(unsigned long) (size=3256) at ../../third_party/libcxx/src/new.cpp:67
#4  0x00007f1b87d6ff99 in std::_LIBCPP_ABI_NAMESPACE::make_unique[abi:v15000]<SkRecorder, decltype(nullptr), SkRect>(decltype(nullptr)&&, SkRect&&) (__args=<error reading variable: Attempt to dereference a generic pointer.>, __args=...)
    at ../../third_party/libcxx/include/__memory/unique_ptr.h:714
#5  0x00007f1b87d6f426 in SkPictureRecorder::SkPictureRecorder() (this=0x561d0b6f0a20)
    at ../../third_party/skia/src/core/SkPictureRecorder.cpp:25
#6  0x00007f1b878635bd in std::_LIBCPP_ABI_NAMESPACE::make_unique[abi:v15000]<SkPictureRecorder>() ()
    at ../../third_party/libcxx/include/__memory/unique_ptr.h:714
#7  0x00007f1b87862d7c in flutter::EmbedderExternalView::EmbedderExternalView(SkISize const&, SkMatrix const&, flutter::EmbedderExternalView::ViewIdentifier, std::_LIBCPP_ABI_NAMESPACE::unique_ptr<flutter::EmbeddedViewParams, std::_LIBCPP_ABI_NAMESPACE::default_delete<flutter::EmbeddedViewParams> >)
    (this=0x7f1b68563a70, frame_size=..., surface_transformation=..., view_identifier=..., params=...)
    at ../../flutter/shell/platform/embedder/embedder_external_view.cc:33
#8  0x00007f1b87862cc0 in flutter::EmbedderExternalView::EmbedderExternalView(SkISize const&, SkMatrix const&)
    (this=0x7f1b68563a70, frame_size=..., surface_transformation=...)
    at ../../flutter/shell/platform/embedder/embedder_external_view.cc:21
#9  0x00007f1b87865dad in std::_LIBCPP_ABI_NAMESPACE::make_unique[abi:v15000]<flutter::EmbedderExternalView, SkISize&, SkMatrix&>(SkISize&, SkMatrix&) (__args=..., __args=...) at ../../third_party/libcxx/include/__memory/unique_ptr.h:714
#10 0x00007f1b87864d1c in flutter::EmbedderExternalViewEmbedder::BeginFrame(SkISize, GrDirectContext*, double, fml::RefPtr<fml::RasterThreadMerger>) (this=0x561d0a1d1340, frame_size=..., context=0x561d0a5abc90, device_pixel_ratio=1, raster_thread_merger=...)
    at ../../flutter/shell/platform/embedder/embedder_external_view_embedder.cc:67
#11 0x00007f1b887a4cf3 in flutter::Rasterizer::DrawToSurfaceUnsafe(flutter::FrameTimingsRecorder&, flutter::LayerTree&)
    (this=0x561d0a1f7ae0, frame_timings_recorder=..., layer_tree=...) at ../../flutter/shell/common/rasterizer.cc:500
#12 0x00007f1b887a2fce in flutter::Rasterizer::DrawToSurface(flutter::FrameTimingsRecorder&, flutter::LayerTree&)
    (this=0x561d0a1f7ae0, frame_timings_recorder=..., layer_tree=...) at ../../flutter/shell/common/rasterizer.cc:471
#13 0x00007f1b887a3fcf in flutter::Rasterizer::DoDraw(std::_LIBCPP_ABI_NAMESPACE::unique_ptr<flutter::FrameTimingsRecorder, std::_LI
```

(Maybe I'm just at a bad engine hash?)

Reply:
Turned out to be incorrect reference usage, fixed now.